File: src/main/java/com/kinvey/sample/micflow/MainActivity.java
package com.kinvey.sample.micflow;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.TextView;

import com.kinvey.android.AsyncAppData;
import com.kinvey.android.Client;
import com.kinvey.android.callback.KinveyListCallback;
import com.kinvey.android.callback.KinveyMICCallback;
import com.kinvey.android.callback.KinveyUserCallback;
import com.kinvey.java.User;

public class MainActivity extends ActionBarActivity {

    private Client kinveyClient;
    private static final String APP_KEY = "kid_Wy7NMiwaTx";
    private static final String APP_SECRET = "18e581bc9c7046a5b1b20ae838105126";
    private static final String redirectURI = "training://";

    private static final String USERNAME = "custom";
    private static final String PASSWORD = "1234";



    private TextView loginStatus;
    private TextView errorView;
    private Button loginWithPage;
    private Button loginAutomated;
    private Button logout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //Create the Kinvey Client with an AppKey and AppSecret
        kinveyClient = new Client.Builder(APP_KEY, APP_SECRET, this).build();

        bindViews();

        updateStatus();

    }

    protected void onResume(){
        super.onResume();
        updateStatus();
    }

    @Override
    public void onNewIntent(Intent intent){
        super.onNewIntent(intent);
        kinveyClient.user().onOAuthCallbackRecieved(intent);
    }

    private void bindViews(){
        loginStatus = (TextView) findViewById(R.id.login_status);
        loginWithPage = (Button) findViewById(R.id.loginpage);
        loginAutomated = (Button) findViewById(R.id.automated);
        logout = (Button) findViewById(R.id.logout);
        errorView = (TextView) findViewById(R.id.login_errors);

        loginWithPage.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                loginWithLoginPage();

            }
        });

        loginAutomated.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                loginAutomated();

            }
        });

        logout.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                logout();

            }
        });
    }


    private void loginWithLoginPage(){
        loading();
        kinveyClient.user().loginWithAuthorizationCodeLoginPage(redirectURI, new KinveyMICCallback() {

            @Override
            public void onSuccess(User user) {
                updateStatus();
            }

            @Override
            public void onFailure(Throwable error) {
                errorView.setText(error.getMessage());
            }

            @Override
            public void onReadyToRender(String myURLToRender) {
                //This renders the login page with the device's default browser
                Uri uri = Uri.parse(myURLToRender);
                Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                startActivity(intent);
            }
        });
    }

    private void loginAutomated(){
        loading();
        kinveyClient.user().setMICApiVersion("v1");
        kinveyClient.user().loginWithAuthorizationCodeAPI(USERNAME, PASSWORD, redirectURI, new KinveyUserCallback() {

            @Override
            public void onSuccess(User user) {
                updateStatus();
            }

            @Override
            public void onFailure(Throwable error) {
                errorView.setText(error.getMessage());
            }
        });

    }

    private void logout(){
        loading();
        try{
            kinveyClient.user().logout().execute();
        }catch(Exception e){
            errorView.setText(e.getMessage());
        }
        updateStatus();
    }

    private void loading(){
        loginStatus.setText("loading...");
    }

    private void updateStatus(){
        if (kinveyClient.user().isUserLoggedIn()){
            loginStatus.setText("User is logged in!");
        }else{
            loginStatus.setText("Not logged in yet!");
        }
        errorView.setText("No Errors!");
    }
}

File: app/src/main/java/com/kinvey/sample/micflow/MainActivity.kt
package com.kinvey.sample.micflow

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.google.api.client.http.HttpTransport
import com.kinvey.android.Client
import com.kinvey.android.Client.Builder
import com.kinvey.android.callback.KinveyUserCallback
import com.kinvey.android.model.User
import com.kinvey.android.store.UserStore
import com.kinvey.java.AbstractClient
import com.kinvey.java.core.KinveyClientCallback
import com.kinvey.java.dto.BaseUser
import com.kinvey.sample.micflow.Constants.APP_CLIENT_ID
import com.kinvey.sample.micflow.Constants.PASSWORD
import com.kinvey.sample.micflow.Constants.REDIRECT_URI
import com.kinvey.sample.micflow.Constants.USERNAME
import com.kinvey.sample.micflow.R.layout
import kotlinx.android.synthetic.main.activity_main.*
import java.util.logging.Level
import java.util.logging.Logger

class MainActivity : AppCompatActivity() {

    private var kinveyClient: Client<User>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_main)
        Logger.getLogger(HttpTransport::class.java.name).level = Level.ALL
        kinveyClient = Builder<User>(this).build()
        bindViews()
        updateStatus()
    }

    override fun onResume() {
        super.onResume()
        updateStatus()
    }

    private fun bindViews() {
        loginPageBtn?.setOnClickListener { loginWithLoginPage() }
        automatedBtn?.setOnClickListener { loginAutomated() }
        logoutBtn?.setOnClickListener { logout() }
    }

    private fun loginWithLoginPage() {
        loading()
        UserStore.presentMICLoginActivity(kinveyClient as Client<*>, APP_CLIENT_ID, REDIRECT_URI,
        object : KinveyUserCallback<User> {
             override fun onSuccess(result: User) {
                 updateStatus()
             }
             override fun onFailure(error: Throwable) {
                 loginErrorsText?.text = error.message
             }
        })
    }

    private fun loginAutomated() {
        loading()
        UserStore.loginWithMIC(kinveyClient as AbstractClient<User>, USERNAME, PASSWORD, APP_CLIENT_ID,
        object: KinveyUserCallback<User> {
             override fun onSuccess(result: User) {
                 updateStatus()
             }
             override fun onFailure(error: Throwable) {
                 loginErrorsText?.text = error.message
             }
        })
    }

    private fun logout() {
        loading()
        UserStore.logout(kinveyClient as AbstractClient<BaseUser>,
            object : KinveyClientCallback<Void?> {
                override fun onSuccess(result: Void?) {
                    updateStatus()
                }
                override fun onFailure(error: Throwable) {
                    loginErrorsText?.text = error.message
                }
            })
        updateStatus()
    }

    private fun loading() {
        loginStatusText?.text = "loading..."
    }

    private fun updateStatus() {
        if (kinveyClient?.clientUser != null) {
            loginStatusText?.text = "User is logged in!"
        } else {
            loginStatusText?.text = "Not logged in yet!"
        }
        loginErrorsText?.text = "No Errors!"
    }
}